### PR TITLE
Support/Enable Operator Lifecycle Manager Webhooks

### LINKED ISSUE #4993: Support webhook in OLM CSV
CSVs now support the definition of Admission and Conversion webhooks. We should look into deploying our existing validating webhook when ECK is installed via OLM so that users have the best experience now matter how they installed the ECK operator. Currently users installing via OLM will get validation errors only in the operator logs and as K8s events which less obvious and inferior to the synchronous feedback the webhook provides.

Reading through https://olm.operatorframework.io/docs/advanced-tasks/adding-admission-and-conversion-webhooks/ one complication is the fact that OLM manages the CA certificates for the webhook and injects the certificates into the operator deployment. This conflicts with our current approach where ECK generates the certificate, however we already have the necessary flags in place to disable this behaviour: 

I believe  `manage-webhook-certs=false` and leaving `webhook-cert-dir={TempDir}/k8s-webhook-server/serving-certs` at its default value should suffice.

### PR DESCRIPTION
**Pull request by naemono**
These changes will enable the eck webhook in OCP clusters with operators managed by the Operator Lifecycle Manager, or any kubernetes cluster with operators managed by the OLM.

closes #4993 

Tested with 

```
❯ oc version
Client Version: 4.9.5
Server Version: 4.7.0
Kubernetes Version: v1.20.0+bd9e442
```

After installing eck-operator using catalog
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: elastic-operators
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: docker.io/mmontg1/operators:1.8.0-rc2
  displayName: Elastic ECK Operator
  publisher: elastic.co
```

and subscription
```
---
apiVersion: v1
kind: Namespace
metadata:
  name: elastic
---
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  annotations:
    olm.providedAPIs: ApmServer.v1.apm.k8s.elastic.co,Elasticsearch.v1.elasticsearch.k8s.elastic.co,Kibana.v1.kibana.k8s.elastic.co
  name: elastic-cloud-eck-group
  namespace: elastic
spec:
  targetNamespaces: []
---
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: elastic-cloud-eck
  namespace: elastic
spec:
  channel: stable
  name: elastic-cloud-eck
  source: elastic-operators
  sourceNamespace: openshift-marketplace
```

We have the following elastic webhooks installed
```
❯ kc get MutatingWebhookConfiguration,ValidatingWebhookConfiguration -A | grep elastic
validatingwebhookconfiguration.admissionregistration.k8s.io/elastic-agent-validation-v1alpha1.k8s.elastic.co-2rw4r   1          18h
validatingwebhookconfiguration.admissionregistration.k8s.io/elastic-apm-validation-v1.k8s.elastic.co-5vlls           1          18h
validatingwebhookconfiguration.admissionregistration.k8s.io/elastic-apm-validation-v1beta1.k8s.elastic.co-wmq2k      1          18h
validatingwebhookconfiguration.admissionregistration.k8s.io/elastic-beat-validation-v1beta1.k8s.elastic.co-8wrln     1          18h
validatingwebhookconfiguration.admissionregistration.k8s.io/elastic-ent-validation-v1.k8s.elastic.co-sdkch           1          18h
validatingwebhookconfiguration.admissionregistration.k8s.io/elastic-ent-validation-v1beta1.k8s.elastic.co-xnbsb      1          18h
validatingwebhookconfiguration.admissionregistration.k8s.io/elastic-es-validation-v1.k8s.elastic.co-8q6wh            1          18h
validatingwebhookconfiguration.admissionregistration.k8s.io/elastic-es-validation-v1beta1.k8s.elastic.co-swz48       1          18h
validatingwebhookconfiguration.admissionregistration.k8s.io/elastic-kb-validation-v1.k8s.elastic.co-mld76            1          18h
validatingwebhookconfiguration.admissionregistration.k8s.io/elastic-kb-validation-v1beta1.k8s.elastic.co-m5zhl       1          18h
```

And the deployment has the necessary arguments
```
❯ kc get deployment -n elastic elastic-operator -o json | jq '.spec.template.spec.containers[0].args'
[
  "manager",
  "--config=/conf/eck.yaml",
  "--manage-webhook-certs=false",
  "--enable-webhook",
  "--distribution-channel=community-operators"
]
```

And an invalid ES instance is rejected by the validating webhook, as expected
```
❯ kc apply -f config/samples/elasticsearch/elasticsearch-invalid-cluster-nomaster.yaml
Error from server (Elasticsearch.elasticsearch.k8s.elastic.co "invalid-cluster-nomaster" is invalid: spec.nodeSets: Required value: Elasticsearch needs to have at least one master node): error when creating "config/samples/elasticsearch/elasticsearch-invalid-cluster-nomaster.yaml": admission webhook "elastic-es-validation-v1.k8s.elastic.co" denied the request: Elasticsearch.elasticsearch.k8s.elastic.co "invalid-cluster-nomaster" is invalid: spec.nodeSets: Required value: Elasticsearch needs to have at least one master node
```

### CODE REVIEW

**Comment by naemono** [comment]
run full pr build

**Comment by naemono** [comment]
*note* that when installing via the generated yaml files, or even helm, you will get one `ValidatingWebhookConfiguration`, with multiple `webhooks`:

```
❯ wget -O - https://download.elastic.co/downloads/eck/1.8.0/operator.yaml 2>/dev/null | yq e '. | select(.kind == "ValidatingWebhookConfiguration") | .webhooks | length' -
10
```

but when installing via OLM, you will get a ValidatingWebhookConfiguration *per* webhook because of the CSV definition which has `webhookdefinitions`, which is an array of definitions, which supports one webhook per definition.

File: hack/operatorhub/main.go [hunk 1]
```diff
@@ -352,19 +389,41 @@ func buildRenderParams(conf *config, packageIndex int, extracts *yamlExtracts) (
...
+		OperatorRBAC:     string(rbac),
+		PackageName:      conf.Packages[packageIndex].PackageName,
+		UbiOnly:          conf.Packages[packageIndex].UbiOnly,
 	}, nil
 }
 
+// validatingWebhookConfigurationToWebhookDefinition converts a standard validating webhook configuration resource
+// to an OLM webhook definition resource.
+func validatingWebhookConfigurationToWebhookDefinition(webhookConfiguration admissionv1.ValidatingWebhookConfiguration) []WebhookDefinition {
+	webhookDefinitions := make([]WebhookDefinition, 0)
```
Comment (pebrc) [inline hack/operatorhub/main.go:409]: Same as above re initialisation

**Review by pebrc — COMMENTED** [review]
LGTM! I was able to test this on my OCP cluster successfully.

File: hack/operatorhub/main.go [hunk 2]
```diff
@@ -246,11 +264,16 @@ func extractYAMLParts(stream io.Reader) (*yamlExtracts, error) {
...
+		return nil, fmt.Errorf("failed to register admissionregistration/v1: %w", err)
+	}
+
 	decoder := scheme.Codecs.UniversalDeserializer()
 	yamlReader := yaml.NewYAMLReader(bufio.NewReader(stream))
 
 	parts := &yamlExtracts{
-		crds: make(map[string]*CRD),
+		crds:             make(map[string]*CRD),
+		operatorWebhooks: make([]admissionv1.ValidatingWebhookConfiguration, 0),
```
[the remark this answers is not in the document]
Comment (naemono) [reply hack/operatorhub/main.go]: true, removed.

**Review by pebrc — APPROVED** [review]
LGTM! So glad that we are finally getting web hooks in OLM 👍

**Pull request merged**